File: 983. Minimum Cost For Tickets/main.py
from typing import List


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        cache = {}

        def dfs(i):
            if i == len(days):
                return 0
            
            if i in cache:
                return cache[i]
            
            res = float('inf')
            for cost, duration in zip(costs, [1, 7, 30]):
                j = i 
                while j < len(days) and days[j] < days[i] + duration:
                    j += 1
                res = min(res, cost + dfs(j))
            
            cache[i] = res
            return res
        
        return dfs(0)
        

File: 983. Minimum Cost For Tickets/main.go
package main

import "math"

func mincostTickets(days []int, costs []int) int {
	durations := []int{1, 7, 30}
	cache := make(map[int]int)

	var dfs func(i int) int
	dfs = func(i int) int {
		if i == len(days) {
			return 0
		}

		if cachedVal, exists := cache[i]; exists {
			return cachedVal
		}

		res := math.MaxInt
		for j := range costs {
			cost := costs[j]
			duration := durations[j]
			t := i
			for t < len(days) && days[t] < days[i]+duration {
				t++
			}
			res = int(math.Min(float64(res), float64(cost+dfs(t))))
		}
		cache[i] = res
		return res
	}

	return dfs(0)
}
